Content.Server/Bed/Sleep/SleepingSystem.cs

```diff
@@ -0,0 +1,198 @@
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
+
+        [Dependency] private readonly IRobustRandom _robustRandom = default!;
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<MobStateComponent, SleepStateChangedEvent>(OnSleepStateChanged);
+            SubscribeLocalEvent<SleepingComponent, DamageChangedEvent>(OnDamageChanged);
+            SubscribeLocalEvent<MobStateComponent, SleepActionEvent>(OnSleepAction);
+            SubscribeLocalEvent<MobStateComponent, WakeActionEvent>(OnWakeAction);
+            SubscribeLocalEvent<SleepingComponent, MobStateChangedEvent>(OnMobStateChanged);
+            SubscribeLocalEvent<SleepingComponent, GetVerbsEvent<AlternativeVerb>>(AddWakeVerb);
+            SubscribeLocalEvent<SleepingComponent, InteractHandEvent>(OnInteractHand);
+            SubscribeLocalEvent<SleepingComponent, ExaminedEvent>(OnExamined);
+            SubscribeLocalEvent<ForcedSleepingComponent, ComponentInit>(OnInit);
+        }
+
+        /// <summary>
+        /// when sleeping component is added or removed, we do some stuff with other components.
+        /// </summary>
+        private void OnSleepStateChanged(EntityUid uid, MobStateComponent component, SleepStateChangedEvent args)
+        {
+            _prototypeManager.TryIndex<InstantActionPrototype>("Wake", out var wakeAction);
+            if (args.FellAsleep)
+            {
+                EnsureComp<StunnedComponent>(uid);
+                EnsureComp<KnockedDownComponent>(uid);
+
+                var emitSound = EnsureComp<SpamEmitSoundComponent>(uid);
+                emitSound.Sound = new SoundCollectionSpecifier("Snores");
+                emitSound.PlayChance = 0.33f;
+                emitSound.RollInterval = 5f;
+                emitSound.PopUp = "sleep-onomatopoeia";
+                emitSound.PitchVariation = 0.2f;
+
+                if (wakeAction != null)
+                {
+                    var wakeInstance = new InstantAction(wakeAction);
+                    wakeInstance.Cooldown = (_gameTiming.CurTime, _gameTiming.CurTime + TimeSpan.FromSeconds(15));
+                    _actionsSystem.AddAction(uid, wakeInstance, null);
+                }
+                return;
+            }
+            if (wakeAction != null)
+                _actionsSystem.RemoveAction(uid, wakeAction);
+
+            RemComp<StunnedComponent>(uid);
+            RemComp<KnockedDownComponent>(uid);
+            RemComp<SpamEmitSoundComponent>(uid);
+        }
+
+        /// <summary>
+        /// Wake up if we take an instance of more than 2 damage.
+        /// </summary>
+        private void OnDamageChanged(EntityUid uid, SleepingComponent component, DamageChangedEvent args)
+        {
+            if (!args.DamageIncreased || args.DamageDelta == null)
+                return;
+
+            if (args.DamageDelta.Total >= 2)
```

Comment:
this value should be on the component (at least so it can get VV'd even if it doesnt serialize). also i think it should be closer to like 5 or so cause there are some poisons that would deal >2 damage

Reply:
If you make it 5 then people can punch mobs with blunt resistance and not wake them up

Reply:
thats fine then